workflow.go

```diff
@@ -0,0 +1,257 @@
+package subscription
+
+import (
+	"errors"
+	"time"
+
+	"go.temporal.io/sdk/workflow"
+)
+
+// This threshold is purposefully short to allow for easier
+// testing and demonstration of Continue-As-New. The Temporal Events in
+// this application are small enough that it's unlikely to hit the event
+// *size* limit, and so in the "real" version of this,
+// EventHistoryThreshold should be much closer to 50k.
+const EventHistoryThreshold = 50
+
+func SubscriptionWorkflow(ctx workflow.Context, subscription SubscriptionInfo) (err error) {
+	ao := workflow.ActivityOptions{
+		StartToCloseTimeout: 10 * time.Second,
+	}
+	ctx = workflow.WithActivityOptions(ctx, ao)
+
+	logger := workflow.GetLogger(ctx)
+	logger.Info("Subscription workflow started", "customer", subscription.Customer, "subscription", subscription)
+
+	err = subscription.trialPeriod(ctx)
+	var continueAsNewErr *workflow.ContinueAsNewError
+	if err != nil && errors.As(err, &continueAsNewErr) {
+		return subscription.maybeContinueAsNew(ctx, err)
+	} else if err != nil {
+		logger.Error("Error while handling trial period", "Error", err)
+		return err
+	}
+
+	err = subscription.billingCycle(ctx)
+	if err != nil && errors.As(err, &continueAsNewErr) {
+		return subscription.maybeContinueAsNew(ctx, err)
+	} else if err != nil {
+		logger.Error("Error while handling billing cycle", "Error", err)
+		return err
+	}
+
+	err = subscription.cancelSubscription(ctx)
+	if err != nil {
+		logger.Error("Canceling subscription failed.", "Error", err)
+		return err
+	}
+
+	logger.Info("Subscription workflow completed.")
+	return nil
+}
+
+func (s *SubscriptionInfo) trialPeriod(ctx workflow.Context) (err error) {
+	logger := workflow.GetLogger(ctx)
+	selector := workflow.NewSelector(ctx)
+	info := workflow.GetInfo(ctx)
+
+	var a *Activities
+	var cancelSub bool
+
+	if !s.TrialPeriodExpiration.IsZero() && s.TrialPeriodExpiration.After(workflow.Now(ctx)) {
+		err = workflow.ExecuteActivity(ctx, a.SendWelcomeEmail, s).Get(ctx, nil)
+		if err != nil {
+			logger.Error("Welcome Email Activity failed.", "Error", err)
+			return err
+		}
+
+		// Start trial period timer, async
+		trialDuration := s.TrialPeriodExpiration.Sub(workflow.Now(ctx))
+		trialTimer := workflow.NewTimer(ctx, trialDuration)
+
+		var timerError error
+		trialOver := false
+		selector.AddFuture(trialTimer, func(f workflow.Future) {
+			logger.Info("Trial has ended. Sending 'trial is over' email, then entering normal billing cycle.")
+			trialOver = true
+			timerError = f.Get(ctx, nil)
+		})
+
+		for !trialOver {
+			selector.
```

Comment:
You should add these signal handlers before the loop. Don't keep adding the same one over and over.